lib/grandpa/grandpa.go

```diff
@@ -38,178 +50,192 @@ func NewService(blockState BlockState, voters []*Voter) (*Service, error) {
-	if m.setID != s.state.setID {
-		return nil, ErrSetIDMismatch
+	return highest, nil
+}
+
+// getPossiblePreVotedBlocks returns blocks with total votes >=2/3 the total number of voters in the map of block hash to block number.
+// if there are no blocks that have >=2/3 direct votes, this function will find predecessors of those blocks that do have >=2/3 votes.
+// note that by voting for a block, all of its predecessor blocks are automatically voted for.
+// thus, if there are no blocks with >=2/3 total votes, but the sum of votes for blocks A and B is >=2/3, then this function returns
+// the first common predecessor of A and B.
+func (s *Service) getPossiblePreVotedBlocks() (map[common.Hash]uint64, error) {
+	// get blocks that were directly voted for
+	votes := s.getDirectVotes()
+	blocks := make(map[common.Hash]uint64)
+
+	// check if any of them have >=2/3 votes
+	for v := range votes {
+		total, err := s.getTotalVotesForBlock(v.hash)
+		if err != nil {
+			return nil, err
+		}
+
+		if total >= uint64(2*len(s.state.voters)/3) {
+			blocks[v.hash] = v.number
+		}
 	}
 
-	// check for equivocation ie. multiple votes within one subround
-	voter, err := s.state.pubkeyToVoter(pk)
-	if err != nil {
-		return nil, err
+	// since we want to select the block with the highest number that has >=2/3 votes,
+	// we can return here since their predecessors won't have a higher number.
+	if len(blocks) != 0 {
+		return blocks, nil
 	}
 
-	vote := NewVote(m.message.hash, m.message.number)
+	// no block has >=2/3 direct votes, check for votes for predecessors recursively
+	var err error
+	va := s.getVotes()
 
-	equivocated := s.checkForEquivocation(voter, vote)
-	if equivocated {
-		return nil, ErrEquivocation
+	for v := range votes {
+		blocks, err = s.getPossiblePreVotedPredecessors(va, v.hash, blocks)
+		if err != nil {
+			return nil, err
+		}
 	}
 
-	err = s.validateVote(vote)
-	if err != nil {
-		return nil, err
-	}
+	return blocks, nil
+}
 
-	s.votes[pk.AsBytes()] = vote
+// getPossiblePreVotedPredecessors recursively searches for predecessors with >=2/3 votes
+// it returns a map of block hash -> number, such that the blocks in the map have >=2/3 votes
+func (s *Service) getPossiblePreVotedPredecessors(votes []Vote, curr common.Hash, prevoted map[common.Hash]uint64) (map[common.Hash]uint64, error) {
+	for _, v := range votes {
+		if v.hash == curr {
+			continue
+		}
 
-	return vote, nil
-}
+		// find common predecessor, check if votes for it is >=2/3 or not
+		pred, err := s.blockState.HighestCommonAncestor(v.hash, curr)
+		if err != nil {
+			return nil, err
+		}
 
-func validateMessageSignature(pk *ed25519.PublicKey, m *VoteMessage) error {
-	msg, err := scale.Encode(&FullVote{
-		stage: m.stage,
-		vote:  NewVote(m.message.hash, m.message.number),
-		round: m.round,
-		setID: m.setID,
-	})
-	if err != nil {
-		return err
-	}
-	ok, err := pk.Verify(msg, m.message.signature[:])
-	if err != nil {
-		return err
-	}
+		if pred == curr {
```

Comment:
this could be a else if condition since its related to the previous statement.

Reply:
what do you mean?

Reply:
oh, I guess you mean 
```
if err != nil { 
} else if pred == curr {
}
```

I prefer not chaining `else`s onto error checks, I don't usually see that happening in go code

Reply:
https://golang.org/doc/effective_go.html#if

looks like if an `if` statement has a return in it, it's better not to chain the else

Reply:
The effective go is saying about another scenario. In this case, a else if is recommended since the var was initialized on the if condition itself, not in the next line like their example:
```
f, err := os.Open(name)
if err != nil {
    return err
}
d, err := f.Stat()
if err != nil {
    f.Close()
    return err
}
codeUsing(f, d)
```

Reply:
it also says right above that code:

> Since error cases tend to end in return statements, the resulting code needs no else statements

I think this is pretty explicit in that there shouldn't be an else statement.